csharp/src/Ice/IncomingResponseFrame.cs

```diff
@@ -13,105 +13,155 @@ public enum ResultType : byte { Success, Failure };
-                    if (InputStream.Size == 0)
-                    {
-                        _payload = ArraySegment<byte>.Empty;
-                    }
-                    else
-                    {
-                        // TODO: for other reply status,return the non-encaps payload.
-                        Debug.Assert(ReplyStatus == ReplyStatus.OK || ReplyStatus == ReplyStatus.UserException);
-
-                        // TODO: works only when Payload called first before reading anything. Need a better version!
-                        // TODO: provide Encoding property
-                        ArraySegment<byte> payload = InputStream.ReadEncapsulation(out Encoding _);
-                        // The payload must included the byte before the encapsulation
-                        // corresponding to the reply status.
-                        Debug.Assert(payload.Array.Length >= payload.Count + 1);
-                        _payload = new ArraySegment<byte>(payload.Array, payload.Offset - 1, payload.Count + 1);
-                    }
+                    var istr = new InputStream(_communicator, Payload, 1);
+                    istr.StartEncapsulation();
+                    T ret = reader(istr);
+                    istr.EndEncapsulation();
+                    return ret;
+                }
+                case ReplyStatus.UserException:
+                {
+                    var istr = new InputStream(_communicator, Payload, 1);
+                    istr.StartEncapsulation();
+                    RemoteException ex = istr.ReadException();
+                    istr.EndEncapsulation();
+                    throw ex;
+                }
+                case ReplyStatus.ObjectNotExistException:
+                case ReplyStatus.FacetNotExistException:
+                case ReplyStatus.OperationNotExistException:
+                {
+                    throw ReadDispatchException();
+                }
+                case ReplyStatus.UnknownException:
+                case ReplyStatus.UnknownLocalException:
+                case ReplyStatus.UnknownUserException:
+                {
+                    throw new UnhandledException(InputStream.ReadString(Payload.Slice(1)), Identity.Empty, "", "");
                 }
-                return _payload.Value;
-            }
-        }
-
-        private ArraySegment<byte>? _payload;
 
-        /// <summary>Take the payload from this response frame. After calling this method, the payload can no longer
-        /// be read.</summary>
-        public ArraySegment<byte> TakePayload()
-        {
-            // TODO: make this method destructive with memory pooling.
-            return Payload;
+                default:
+                {
+                    Debug.Assert(false);
+                    throw new InvalidDataException(
+                        $"received ice1 response frame with unknown reply status `{ReplyStatus}'");
+                }
```

Comment:
Right, that is why I added the Assert, the compiler complains that not all code paths return a value without this.

Reply:
How come it doesn't complain for the ReadVoidReturnValue implementation where we don't have a default case in the switch?

Reply:
Because it is void, the complaint is about not all code paths returning a value, I think I will fix this with suggested `ReadExceptionReturnValue` used by both